File: Day47/Part2/minion_network.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

struct Edge {
    int u, v;
    int cost;
    bool is_corrupted;
    int actual_cost;

    bool operator<(const Edge& other) const {
        return actual_cost < other.actual_cost;
    }
};

struct UnionFind {
    vector<int> parent;
    vector<int> rank;
    int count;

    UnionFind(int n) {
        parent.resize(n);
        iota(parent.begin(), parent.end(), 0);
        rank.assign(n, 0);
        count = n;
    }

    int find(int u) {
        if (parent[u] != u) {
            parent[u] = find(parent[u]);
        }
        return parent[u];
    }

    bool unite(int u, int v) {
        int root_u = find(u);
        int root_v = find(v);

        if (root_u != root_v) {
            if (rank[root_u] > rank[root_v]) {
                parent[root_v] = root_u;
            } else if (rank[root_u] < rank[root_v]) {
                parent[root_u] = root_v;
            } else {
                parent[root_v] = root_u;
                rank[root_u]++;
            }
            count--;
            return true;
        }
        return false;
    }
};

int solveMinionNetwork(int n, vector<Edge>& connections) {
    // Calculate actual costs
    for (auto& edge : connections) {
        edge.actual_cost = edge.is_corrupted ? edge.cost * 2 : edge.cost;
    }

    // Sort edges
    sort(connections.begin(), connections.end());

    UnionFind uf(n);
    int total_mana = 0;
    int edges_count = 0;

    // cout << "Processing " << connections.size() << " potential connections for " << n << " outposts..." << endl;

    for (const auto& edge : connections) {
        if (uf.unite(edge.u, edge.v)) {
            // cout << "  Connecting " << edge.u << " <-> " << edge.v << " with cost " << edge.actual_cost << endl;
            total_mana += edge.actual_cost;
            edges_count++;
        }
    }

    if (uf.count == 1) {
        return total_mana;
    } else {
        return -1;
    }
}

int main() {
    int N = 4;
    // {u, v, cost, is_corrupted, actual_cost (placeholder)}
    vector<Edge> connections = {
        {0, 1, 10, false, 0},
        {1, 2, 20, true, 0},
        {2, 3, 30, false, 0},
        {0, 3, 50, false, 0},
        {1, 3, 5, false, 0}
    };

    int result = solveMinionNetwork(N, connections);
    cout << "Minimum Mana Cost: " << result << endl;

    return 0;
}

File: Day47/Part2/minion_network.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n
        self.count = n  # Number of disjoint sets

    def find(self, u):
        if self.parent[u] != u:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]

    def union(self, u, v):
        root_u = self.find(u)
        root_v = self.find(v)

        if root_u != root_v:
            if self.rank[root_u] > self.rank[root_v]:
                self.parent[root_v] = root_u
            elif self.rank[root_u] < self.rank[root_v]:
                self.parent[root_u] = root_v
            else:
                self.parent[root_v] = root_u
                self.rank[root_u] += 1
            self.count -= 1
            return True
        return False

def solve_minion_network(n, connections):
    # Calculate actual costs and prepare edges
    # connections: [u, v, cost, is_corrupted]
    edges = []
    for u, v, cost, is_corrupted in connections:
        actual_cost = cost * 2 if is_corrupted else cost
        edges.append((actual_cost, u, v))
    
    # Sort edges by cost
    edges.sort()
    
    uf = UnionFind(n)
    total_mana = 0
    edges_used = 0
    
    # print(f"Processing {len(edges)} potential connections for {n} outposts...")
    
    for cost, u, v in edges:
        if uf.union(u, v):
            # print(f"  Connecting {u} <-> {v} with cost {cost}")
            total_mana += cost
            edges_used += 1
            
    if uf.count == 1:
        return total_mana
    else:
        return -1

# Test Case
if __name__ == "__main__":
    N = 4
    # [u, v, cost, is_corrupted]
    Connections = [
        [0, 1, 10, 0],
        [1, 2, 20, 1], # Actual 40
        [2, 3, 30, 0],
        [0, 3, 50, 0],
        [1, 3, 5, 0]
    ]
    
    result = solve_minion_network(N, Connections)
    print(f"Minimum Mana Cost: {result}")
